src/Physics/Coherent/EventGen/COHGammaKinematicsGenerator.cxx

```diff
@@ -0,0 +1,426 @@
+  // neutrino energy in the LAB rather than in the hit nucleon rest frame.
+
+  const InitialState & init_state = interaction->InitState();
+  double E = init_state.ProbeE(kRfLab);
+  return E;
+}
+//___________________________________________________________________________
+void COHGammaKinematicsGenerator::throwOnTooManyIterations(unsigned int iters,
+                                                      GHepRecord* evrec) const
+{
+  LOG("COHKinematics", pWARN)
+    << "*** Could not select valid kinematics after "
+    << iters << " iterations";
+  evrec->EventFlags()->SetBitNumber(kKineGenErr, true);
+  genie::exceptions::EVGThreadException exception;
+  exception.SetReason("Couldn't select kinematics");
+  exception.SwitchOnFastForward();
+  throw exception;
+}
+//___________________________________________________________________________
+void COHGammaKinematicsGenerator::Configure(const Registry & config)
+{
+  Algorithm::Configure(config);
+  this->LoadConfig();
+}
+//____________________________________________________________________________
+void COHGammaKinematicsGenerator::Configure(string config)
+{
+  Algorithm::Configure(config);
+  this->LoadConfig();
+}
+//____________________________________________________________________________
+void COHGammaKinematicsGenerator::LoadConfig(void)
+{
+
+  bool error = false ;
+
+  GetParamDef( "tPhaseSpace", ftPhaseSpace, false ) ;
+  
+  
+  //-- max xsec safety factor (for rejection method) and min cached energy
+  GetParamDef( "MaxXSec-SafetyFactor", fSafetyFactor, 1.6 ) ;
+  GetParamDef( "Cache-MinEnergy", fEMin,  -1.0 ) ;
+
+  //-- Generate kinematics uniformly over allowed phase space and compute
+  //   an event weight?
+  GetParamDef( "UniformOverPhaseSpace", fGenerateUniformly, false ) ;
+
+  //-- Maximum allowed fractional cross section deviation from maxim cross
+  //   section used in rejection method
+  GetParamDef( "MaxXSec-DiffTolerance", fMaxXSecDiffTolerance, 999999. ) ;
+  assert(fMaxXSecDiffTolerance>=0);
+
+  int max_iter ;
+  GetParamDef( "MaxIterations", max_iter, (int) kRjMaxIterations ) ;
+  if ( max_iter > 0 ) {
+    fMaxIterations = max_iter ;
+  }
+
+  const Algorithm * temp = SubAlg( "IntegrationLimits" ) ;
+  fGammaLimits = dynamic_cast<const COHGammaIntegrationLimits *>( temp ) ;
+  if (! fGammaLimits ) {
+    LOG( "COHGammaKinematicsGenerator", pERROR ) << "Gamma integration limits subalgo failed to load" ;
+    error = true ;
+  }
+
+  std::vector<int> scan_points ;
+  GetParamVect( "MinimScanPoint", scan_points ) ;
+  if ( scan_points.size() < fMinimScanPoints.size() ) {
+    LOG( "COHGammaKinematicsGenerator", pERROR ) << "Not enough information for phase space scan" ;
+    error = true ;
+  }
+		
+  for ( unsigned int i = 0 ; i < fMinimScanPoints.size() ; ++i ) {
+    fMinimScanPoints[i] = std::max( 1, scan_points[i] ) ;
+  }
+  
+  if ( error ) {
+    LOG( "COHGammaKinematicsGenerator", pFATAL ) << "Invalid configuration. Exiting" ;
```

Comment:
Yes, right now all the new algorithms I'm adding are failing in a similar way. When (and if) I will have time, I plan to change the prototype of `Configure` functions so that a boolean is returned and the exit can be decided at the high level. But I think this is far beyond the scope of this PR.